### PR TITLE
better handling of host lists

### LINKED ISSUE #17762: Using an empty inventory group no longer noops and instead results in error.
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ansible
##### ANSIBLE VERSION

Old "good/desired" behavior observed with ansible 2.0.0.2 as installed via `pip install ansible==2.0.0.2`.

New "bad/undesired" behavior observed with ansible 2.1.1.0 as installed via `pip install ansible==2.1.1.0`.
##### CONFIGURATION

No changes made.
##### OS / ENVIRONMENT

Ubuntu Xenial 16.04
##### SUMMARY

If you run ansible against an empty inventory group this now results in an error and failure. Old behavior was to noop and do nothing.
##### STEPS TO REPRODUCE

```
# Install ansible 2.1.1.0
virtualenv /tmp/ansible
# note the paramiko version is fixed to avoid needing system packages.
/tmp/ansible/bin/pip install paramiko==1.16.0 ansible==2.1.1.0
export ANSIBLE=/tmp/ansible/bin/ansible
echo "[primary]" > /tmp/inventory
echo "localhost ansible_connection=local host_counter=1" >> /tmp/inventory
echo "[subnodes]" >> /tmp/inventory
$ANSIBLE subnodes -f 5 -i "/tmp/inventory" -m copy -a "src='/tmp/inventory' dest='/tmp/foo'"
```
##### EXPECTED RESULTS

I expect to continue to have ansible noop on empty groups.
##### ACTUAL RESULTS

`ERROR! Specified hosts and/or --limit does not match any hosts`

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE

- Bugfix Pull Request
##### COMPONENT NAME

adhoc
##### ANSIBLE VERSION

```
2.2
```
##### SUMMARY

fixes #17762

### CODE REVIEW

**Review by emonty — APPROVED** [review]
lgtm

**Comment by abadger** [comment]
+1

File: lib/ansible/cli/adhoc.py [hunk 1]
```diff
@@ -139,9 +139,12 @@ def run(self):
 
         inventory.subset(self.options.subset)
         hosts = inventory.list_hosts(pattern)
-        if len(hosts) == 0 and no_hosts is False:
-            # Invalid limit
-            raise AnsibleError("Specified hosts and/or --limit does not match any hosts")
+        if len(hosts) == 0:
+            if no_hosts is False and  self.options.subset:
```
Comment (resmo) [inline lib/ansible/cli/adhoc.py]: double space between `and` and `self.options.subset`

**Comment by resmo** [comment]
+1

**Review by Shaps — APPROVED** [review]
+1

**Pull request merged**